### ISSUE TITLE
Using .with<T>(instance)

### ISSUE THREAD

**Issue by jlhorner1974** [ybainier/Hypodermic#24]
I'm trying to set up a call to .with<T>() when registering a type with an existing instance of an object, following the sample on the Hypodermic github site, as shown below. The .with<T>() executes without incident, but it does not wire up the dependency in the parent object, which is left empty.  According to the docs, it seems this should work. Is there something I'm doing wrong?

class IFoo
{
public:
	virtual void Bar() = 0;
};

class Foo : public IFoo
{
public:
	virtual void Bar() override {};
};

class IBaz { };

class Baz : public IBaz
{
public:
	Baz(std::shared_ptr<IFoo> foo) : m_pFoo(foo) {};
	std::shared_ptr<IFoo> m_pFoo;
};

Hypodermic::ContainerBuilder builder;
auto foo = std::make_shared<Foo>();
**builder.registerType<Baz>() .with<IFoo>(foo) .as<IBaz>();**
builder.build();

auto pBaz = pContainer->resolve<Baz>();

**Comment by ybainier** [comment]
Hi, your example do not compile, you are missing the template parameters.

Assuming you are trying to resolve `Baz` which should be injected a `IFoo`, and that you want to _resolve_ `Baz` _as_ a `IBaz`:

```cpp
ContainerBuilder builder;

auto foo = std::make_shared< Foo >();

builder.registerType< Baz >()
           .with< IFoo >(foo)
           .as< IBaz >();

auto container = builder.build();

auto baz = container->resolve< IBaz >();
```

But you could write it that way too:

```cpp
ContainerBuilder builder;

auto foo = std::make_shared< Foo >();

builder.registerInstance(foo).as< IFoo >();
builder.registerType< Baz >().as< IBaz >();

auto container = builder.build();

auto baz = container->resolve< IBaz >();
```

Is this solving your problem?

**Comment by jlhorner1974** [comment]
Ah, I see now (I did have the template types in my post, but I didn't mark them text as code, so they got stripped when I submitted the post. 

I see my problem now, and I feel stupid. In my resolve call, I had a typo where I was specifying the type to be resolved, I has passed in Baz instead of IBaz, so of course, it wasn't working.

I fixed my mistake and all works as it should. Thank you for your help.
